### PR TITLE
Sign in with GitHub

### LINKED ISSUE #127: Log in with GitHub
We'd like to be able to log in with our GitHub account.

Later, in different issues, we'll use this to like notebooks and add comments.

### PR DESCRIPTION
**Pull request by fabrithedev**
Resolves #127 

### Features Implemented:
- **GitHub Login**: Users can now sign in using their GitHub accounts.
- **Login & Logout Functionalities**: Implemented login and logout functionalities.
- **User Data Retrieval Plug**: Introduced a plug to fetch the logged-in user's data.
- **User Model Updates**: Added new requested fields to the User model to accommodate additional information.

### Details:
This pull request introduces the ability for users to sign in using their GitHub credentials. The implementation includes the necessary login and logout functionalities, ensuring a seamless user experience. Additionally, a new plug has been created to retrieve the data of the currently logged-in user, which is essential for personalized interactions within the application.

The User model has been updated to include new fields as requested, which will store relevant information obtained from the GitHub profile upon authentication.

### How to Test:
1. Navigate to the sign-in page.
2. Click on the "Sign in with GitHub" button.
3. Follow the prompts to authorize the application with GitHub.
4. Once signed in, verify that the new user fields are populated with your GitHub information.
5. Test the logout functionality to ensure the session is cleared correctly.

Please review the changes and provide your feedback.

Thank you!

P.S: used IA to help me write this PR message, sorry for laziness! lmao

### CODE REVIEW

File: config/secret [hunk 1]
```diff
@@ -3,3 +3,8 @@
 import Config
 
 config :notesclub, :github_api_key, "GITHUB_API_KEY"
+
+config :ueberauth, Ueberauth.Strategy.Github.OAuth,
+  client_id: "GITHUB_OAUTH_CLIENT_ID",
+  client_secret: "GITHUB_OAUTH_CLIENT_SECRET"
+
```
Comment (hectorperez) [inline config/secret]: The last line has two spaces. GitHub complains unless there is an empty line at the end.

File: lib/notesclub_web/controllers/auth_controller.ex [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+          {:error, _changeset} ->
+            conn
+            |> put_session(:user_id, nil)
+            |> put_flash(:error, "There was an issue creating your account.")
+            |> redirect(to: "/")
+        end
+
+      user ->
+        Accounts.update_user(user, %{last_login_at: DateTime.utc_now()})
```
Comment (hectorperez) [inline lib/notesclub_web/controllers/auth_controller.ex]: I suppose GitHub provides all params every time we log in, does it?
Then, we could update them (not tested):

```suggestion
  def callback(%{assigns: %{ueberauth_auth: auth}} = conn, _params) do
    user_params = %{
      username: auth.info.nickname,
      github_id: auth.uid,
      name: auth.info.name,
      avatar_url: auth.info.urls.avatar_url,
      bio: auth.extra.raw_info.user["bio"],
      email: auth.info.email,
      location: auth.info.location,
      followers_count: auth.extra.raw_info.user["followers"],
      last_login_at: DateTime.utc_now()
    }
    
    case Accounts.get_by_github_id(auth.uid) do
      nil ->
        case Accounts.create_user(user_params) do
          {:ok, user} ->
            conn
            |> put_session(:user_id, user.id)
            |> redirect(to: "/")

          {:error, _changeset} ->
            conn
            |> put_session(:user_id, nil)
            |> put_flash(:error, "There was an issue creating your account.")
            |> redirect(to: "/")
        end

      user ->
        Accounts.update_user(user, user_params)
```

**Review by hectorperez — COMMENTED** [review]
This is brilliant @FabriDamazio! 🚀 

Two minor suggestions below!
Thanks!

File: lib/notesclub_web/controllers/auth_controller.ex (hunk 2 above)
Comment (fabrithedev) [reply lib/notesclub_web/controllers/auth_controller.ex]: Yes, we can update on every login. Good idea to keep all information updated. I will do it.

**Comment by hectorperez** [comment]
By the way @FabriDamazio, do we need something like this in `runtime.exs` for production?
```elixir
if config_env() == :prod do
    ...
    config :ueberauth, Ueberauth.Strategy.Github.OAuth,
        client_id: System.get_env("GITHUB_AUTH_CLIENT_ID"),
        client_secret: System.get_env("GITHUB_AUTH_SECRET")
end
```

**Comment by fabrithedev** [comment]
> By the way @FabriDamazio, do we need something like this in `runtime.exs` for production?
> 
> ```elixir
> if config_env() == :prod do
>     ...
>     config :ueberauth, Ueberauth.Strategy.Github.OAuth,
>         client_id: System.get_env("GITHUB_AUTH_CLIENT_ID"),
>         client_secret: System.get_env("GITHUB_AUTH_SECRET")
> end
> ```

yes, I'll add it.

File: config/secret (hunk 1 above)
Comment (fabrithedev) [reply config/secret]: 7ebb67cecbdfe064f4db480cf256dabe5af8a84b

File: lib/notesclub_web/controllers/auth_controller.ex (hunk 2 above)
Comment (fabrithedev) [reply lib/notesclub_web/controllers/auth_controller.ex]: 5667d9453c8bc049c59037a97fa9eb357e6a7f5d

**Comment by hectorperez** [comment]
All looks good, merging, thanks @FabriDamazio!

**Pull request merged**